### PR TITLE
feat: coalesce view-session events for Email notification

### PR DESCRIPTION
**Pull request by xiez**

### CODE REVIEW

File: backend/automations/emails.py [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+                        status=AutomationDelivery.Status.PENDING
+                    ).update(status=AutomationDelivery.Status.SUCCESS, delivered_at=timezone.now())
+                    
+                    if updated_count > 0:
+                        should_send = True
+                else:
+                    logger.debug(
+                        'Automation delivery debounced: visitor is still active (%s seconds since last event). view_session_id=%s',
+                        int(time_since_latest),
+                        view_session_id
```
Comment (xiez) [inline backend/automations/emails.py]: **Concurrency / Race Condition Defect**

Using `.update()` for Optimistic Concurrency Control with `if updated_count > 0:` is flawed for batch operations that can grow.

If two workers run concurrently for this session:
1. Worker A queries and finds `[Delivery 1]`.
2. Worker B queries a split-second later and finds `[Delivery 1, Delivery 2]`.
3. Worker A runs `.update()`, updates `Delivery 1`, gets `updated_count = 1`, and sends the email containing `Delivery 1`.
4. Worker B runs `.update()`. Since `Delivery 1` is already `SUCCESS`, it only updates `Delivery 2`, getting `updated_count = 1`.
5. Because `updated_count > 0` is `True`, Worker B will also dispatch an email containing its original list: `[Delivery 1, Delivery 2]`.

The user will receive duplicate emails with overlapping events.

To prevent this, follow the spirit of **Rule 1** (`select_for_update` for concurrency-sensitive reads). Use `select_for_update()` inside `transaction.atomic()` when aggregating/coalescing multiple rows, instead of relying on a flawed `updated_count` check.

File: backend/automations/emails.py [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+
+def handle_email_delivery(delivery: AutomationDelivery):
+    if delivery.status != AutomationDelivery.Status.PENDING:
+        logger.info('Automation delivery already processed: delivery_id=%s status=%s', delivery.id, delivery.status)
+        return
+
+    owner = delivery.rule.created_by
+    recipient_email = owner.email
+    if not recipient_email:
```
Comment (xiez) [inline backend/automations/emails.py]: **Broken Retry Mechanism Defect**

This strict guard clause breaks the exponential retry behavior for all email deliveries.

When a transient SMTP failure occurs, `_mark_failure_and_retry` sets the delivery's status to `FAILED` and schedules a retry task. However, when that retry task wakes up, it will enter this function, see that the status is no longer `PENDING`, log a message, and immediately abort. The retry is silently discarded, and the email is permanently lost.

Furthermore, even if you bypassed this guard, the coalescing query below (`AutomationDelivery.objects.filter(status=...PENDING)`) only searches for `PENDING` deliveries, meaning the `FAILED` ones would still be completely ignored by the aggregator.

To fix this, you must either explicitly allow `FAILED` states to be processed and aggregated, or ensure that the retry handler reverts transient failures back to the `PENDING` state.

File: backend/automations/emails.py [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+    if not receive_email_notification:
+        logger.info('Skipping email alert: notifications disabled for link %s', share_link_id)
+        return
+
+    # Build coalesced context
+    dr_name = payload.get('dataroom_name')
+    doc_name = payload.get('document_name')
+    target_name = dr_name or doc_name or "Shared Item"
+    target_type = 'dataroom' if dr_name else 'document'
```
Comment (xiez) [inline backend/automations/emails.py:127]: 🔴 **[Critical] Duplicate notification check runs AFTER rows are already committed to SUCCESS**

By the time `_send_coalesced_session_email` is called, the deliveries have already been bulk-updated to `SUCCESS` atomically inside `handle_email_delivery`. If `receive_email_notification` is `False` here, the function silently returns — but those rows are durably committed as delivered with no audit trail.

This guard is redundant because `handle_email_delivery` already performs the same `receive_email_notification` check (before the session branch) and calls `_mark_success_direct(..., 'Skipped: ...')`. Remove this duplicate check from `_send_coalesced_session_email` entirely.

File: backend/automations/emails.py [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+        # Revert success status on failure so it can retry
+        with transaction.atomic():
+            for d in pending_deliveries:
+                _mark_failure_and_retry(d, f'Failed to send email: {e}')
+
+
+def _send_standard_fallback_email(owner, recipient_email: str, delivery: AutomationDelivery):
+    from .tasks import _target_description, _build_event_sentence, _display_actor, _mark_success_direct, _mark_failure_and_retry
+    
+    payload = delivery.payload or {}
```
Comment (xiez) [inline backend/automations/emails.py:231]: 🔴 **[Critical] Email sent outside `transaction.atomic()` but rows committed inside — rollback on failure is broken**

The atomic block (L389–411) commits all deliveries to `SUCCESS` *before* this email call. If `send_mail` raises here, `_mark_failure_and_retry` attempts to revert already-durably-committed `SUCCESS` rows, creating conflicting state.

The original `tasks.py` approach was safer: call `_mark_success_direct` only after a successful send. Consider a two-phase approach: claim rows with a `PROCESSING` sentinel status inside the transaction, then mark `SUCCESS` or `FAILED` outside based on the actual send result.

File: backend/automations/emails.py [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+            duration_str = f"{total_duration}s"
+            
+        if num_pages > 1:
+            stats[doc_name] = f"{duration_str}, {completion}% read"
+        else:
+            stats[doc_name] = f"{duration_str}"
+        
+    return stats
+
+
```
Comment (xiez) [inline backend/automations/emails.py:98]: 🔴 **[Critical] Doc name used as dict key — silently overwrites stats for duplicate document names**

If a dataroom contains two documents with the same name (e.g. two `"Report.pdf"` files in different folders), the second document's stats silently overwrite the first's, producing incorrect read percentages and durations in the email.

Use `doc.id` as the internal key and convert to name only at serialization time:
```python
stats[doc.id] = {'name': doc_name, 'info': f"{duration_str}, {completion}% read"}
```

File: backend/automations/emails.py [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+            return f'Malware scanning failed for file "{uploaded_file_name}"'
+        return "Malware scanning failed for file"
+
+    return f'Automation event "{event_type}" was triggered'
+
+
+def _get_session_completion_stats(view_session_id: str, since_datetime=None) -> dict:
+    """
+    Query and compile document reading statistics associated with a view session.
+
```
Comment (xiez) [inline backend/automations/emails.py:41]: 🟡 **[Medium] Dead code — `_build_action_description` is never called**

This function is defined but not referenced anywhere in the module — not in `_send_coalesced_session_email`, not in any template context. Either wire it into the coalesced email digest as a per-event activity log, or remove it to avoid future confusion.

File: backend/automations/emails.py [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+        'viewer_email': viewer_email,
+        'viewed_at': viewed_at,
+        'ip_address': payload.get('visitor_ip'),
+        'location': "Unknown Location",
+        'os': os_name,
+        'browser': browser_name,
+        'activity_stats': [],
+        'activity_stats_txt': [],
+        'show_details_url': _build_show_details_url(payload),
+    }
```
Comment (xiez) [inline backend/automations/emails.py:275]: 🟡 **[Medium] Hardcoded `"Unknown Location"` ignores available payload data**

The `payload` here may contain `visitor_city` and `visitor_country` (they are populated by `_build_visitor_context` in `sharelinks/views.py`). The coalesced path correctly constructs location from the payload, but this standard fallback path always shows `"Unknown Location"` regardless.

```python
city = payload.get('visitor_city')
country = payload.get('visitor_country')
location = f"{city}, {country}" if city and country else "Unknown Location"
```

File: backend/automations/emails.py [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+                recipient_email=recipient_email,
+                view_session_id=view_session_id,
+                pending_deliveries=pending_deliveries,
+                first_delivery=delivery
+            )
+        return
+
+    # Standard fallback for non-session deliveries (e.g., file request upload alerts)
+    _send_standard_fallback_email(
+        owner=owner,
```
Comment (xiez) [inline backend/automations/emails.py:394]: 🟡 **[Medium] Triggering delivery may not be the earliest — use `session_deliveries[0]` instead**

`delivery` here is whichever delivery triggered *this* Celery task, which could be any delivery in the coalesced batch (not necessarily the earliest). `_send_coalesced_session_email` uses `first_delivery.payload` for the email context (target name, share link, viewer, `event_datetime`). If this is not the earliest delivery, the timestamp shown in the email will be incorrect.

Fix: pass the chronologically first delivery from the ordered query:
```python
first_delivery=session_deliveries[0]  # ordered by created_at
```

File: backend/backend/utils.py [hunk 9]
```diff
@@ -79,3 +79,52 @@ def get_unique_name(model_class, original_name: str, filter_kwargs: dict, has_ex
...
+    elif "safari/" in ua_lower and "applewebkit/" in ua_lower:
+        browser_name = "Safari"
+    elif "firefox/" in ua_lower:
+        browser_name = "Firefox"
+    elif "msie" in ua_lower or "trident/" in ua_lower:
+        browser_name = "Internet Explorer"
+    
+    # Fallback to display the raw string if it's brief
+    if os_name == "Unknown OS" and browser_name == "Unknown Browser" and len(ua_string) < 30:
+        return ua_string, ""
```
Comment (xiez) [inline backend/backend/utils.py]: 🟡 **[Medium] Empty string for browser leaks into email template**

When the UA string is short and unrecognised, this returns `(ua_string, "")`. The empty string `""` for browser will render as a blank `<span class="detail-value">` in the HTML email.

Fix: use a proper fallback string:
```python
return ua_string, "Unknown Browser"
```

File: backend/tests/automations/test_tasks.py [hunk 10]
```diff
@@ -495,3 +495,140 @@ def test_deliver_task_email_destination_skipped(mock_send_mail, user, share_link
...
+    AutomationDelivery.objects.filter(status=AutomationDelivery.Status.PENDING).update(
+        status=AutomationDelivery.Status.SUCCESS,
+        delivered_at=timezone.now()
+    )
+
+    # Now execute the task (simulating Task 2 running after Task 1 succeeded)
+    deliver_automation_delivery_task(str(delivery2.id))
+
+    # Assert that no email was sent because CAS update matching status=PENDING returned 0 updated rows
+    mock_send_mail.assert_not_called()
```
Comment (xiez) [inline backend/tests/automations/test_tasks.py:633]: 🟡 **[Medium] No test for the debounce "visitor still active" path**

Tests cover the happy path and the CAS race condition, but there is no test that verifies a delivery is *not* sent when the latest delivery is less than 60 seconds old (visitor still active). Per project TDD policy, acceptance-criteria boundary conditions require a test.

Suggested test name: `test_deliver_task_email_debounced_when_visitor_still_active`

File: backend/automations/services.py [hunk 11]
```diff
@@ -118,7 +118,10 @@ def dispatch_automation_event(event_type: str, payload: dict) -> int:
             # Queue delivery execution immediately after creating the log row.
             # Import locally to avoid circular imports (tasks imports this service).
             from .tasks import deliver_automation_delivery_task
-            deliver_automation_delivery_task.delay(str(delivery.id))
+            if destination.destination_type == 'email':
+                deliver_automation_delivery_task.apply_async(args=[str(delivery.id)], countdown=60)
```
Comment (xiez) [inline backend/automations/services.py]: 🟢 **[Low] Hardcoded `countdown=60` duplicates `idle_threshold_seconds = 60` in `emails.py`**

These two values must stay in sync — if the idle threshold in the email handler is changed (e.g. to 90s) without updating this countdown, the Celery task will fire before the debounce window closes and the email will never be sent.

Extract to a shared constant (e.g. in `emails.py` or a shared `constants.py`) and import it here:
```python
from .emails import EMAIL_DEBOUNCE_SECONDS
deliver_automation_delivery_task.apply_async(args=[str(delivery.id)], countdown=EMAIL_DEBOUNCE_SECONDS)
```

**Review by xiez — COMMENTED** [review]
## Code Review — PR #267: Coalesce view-session events for Email notification

Overall this is a well-structured refactor with a solid concurrency model. The coalescing logic, `select_for_update()` locking, and CAS update pattern are all sound approaches. However there are a few critical correctness issues that need to be addressed before merging.

### 🔴 Critical (must fix)
1. **Silent data integrity bug** — `receive_email_notification` is checked *after* rows are already committed to `SUCCESS`. Deliveries are marked as delivered even when no email is sent.
2. **Broken rollback on send failure** — rows are committed to `SUCCESS` inside the transaction *before* the email is sent. `_mark_failure_and_retry` conflicts with the already-durable `SUCCESS` state.
3. **Duplicate doc name key collision** — `_get_session_completion_stats` uses `doc_name` as the dict key, silently overwriting stats when two documents share the same name.

### 🟡 Medium (should fix)
4. `_build_action_description` is dead code — never called.
5. `_send_standard_fallback_email` hardcodes `"Unknown Location"` even when city/country are available in the payload.
6. `first_delivery=delivery` may not be the chronologically earliest — use `session_deliveries[0]`.
7. `parse_user_agent` fallback returns empty string `""` for browser, rendering a blank field in the email template.
8. Missing test for the debounce "visitor still active" boundary condition.

### 🟢 Low (nit)
9. `countdown=60` in `services.py` duplicates `idle_threshold_seconds = 60` in `emails.py`. Extract to a shared constant to prevent drift.
